pageserver/src/tenant/storage_layer/delta_layer.rs

```diff
@@ -414,16 +424,29 @@ impl DeltaLayerWriterInner {
         // Create the file initially with a temporary filename. We don't know
         // the end key yet, so we cannot form the final filename yet. We will
         // rename it when we're done.
-        //
-        // Note: This overwrites any existing file. There shouldn't be any.
-        // FIXME: throw an error instead?
         let path =
             DeltaLayer::temp_path_for(conf, &tenant_shard_id, &timeline_id, key_start, &lsn_range);
+        let file = TempVirtualFile::new(
+            VirtualFile::open_with_options_v2(
+                &path,
+                virtual_file::OpenOptions::new()
+                    .create_new(true)
+                    .write(true),
+                ctx,
+            )
+            .await?,
+            gate.enter()?,
+        );
 
-        let mut file = TempVirtualFile::new(VirtualFile::create(&path, ctx).await?, gate.enter()?);
-        // make room for the header block
-        file.seek(SeekFrom::Start(PAGE_SZ as u64)).await?;
-        let blob_writer = BlobWriter::new(file, PAGE_SZ as u64, gate, cancel, ctx);
+        // Start at PAGE_SZ, make room for the header block
+        let blob_writer = BlobWriter::new(
+            file,
+            PAGE_SZ as u64,
```

Comment:
This is probably implicit for people familiar with this code, but the contract here wasn't clear to me, and it took a bit of digging to find out. Could you please document that this will in effect seek to the start offset via `write_at`, and extend the file with zeroes if the offset is beyond the end? Concretely, passing a file opened in append mode won't do what the caller might expect (it's kind of given by `start_offset`, but it's not clear whether `start_offset` will actively seek, or whether it's passing in the current file position).

Reply:
I removed all the implicitly-positioned VirtualFile APIs, so, all writes ever done to a VirtualFile carry the offfset.

So, the concern is moot.

Haven't thought about O_APPEND, but we don't use that anyway.

I'll see if I can get some sensible documentation in place.

Reply:
https://github.com/neondatabase/neon/pull/11558/commits/91001788f835343495f8b6dfe9416bc6d46aa6a3